src/aggregation/Aggregator.ts

```diff
@@ -243,3 +244,116 @@ export class LongSumAggregator extends AbstractAggregator<Long> {
         output.writeLong(Long.ZERO);
     }
 }
+
+/** @internal */
+export class DistinctValuesAggregator<R> extends AbstractAggregator<Set<R>> {
+
+    classId = AggregatorConstants.DISTINCT;
+
+    readData(input: DataInput): any {
+        this.attributePath = input.readString();
+        const count = input.readInt();
+        for (let i = 0; i < count; i++) {
+            input.readObject();
+        }
+    }
+
+    writeData(output: DataOutput): void {
+        output.writeString(this.attributePath);
+        output.writeInt(0);
+    }
+}
+
+/** @internal */
+export class CanonicalizingHashSet<R> implements IdentifiedDataSerializable, Set<R> {
+
+    classId = AggregatorConstants.CANONICALIZING_SET;
+    factoryId = AggregatorConstants.AGGREGATOR_FACTORY_ID;
+    private _values = new Set<R>();
+
+    readData(input: DataInput): void {
+        const count = input.readInt();
+        for (let i = 0; i < count; i++) {
+            const element = input.readObject();
+            this._values.add(element);
+        }
+    }
+
+    writeData(output: DataOutput): void {
+        output.writeInt(this._values.size);
+        for (const element of this._values) {
+            output.writeObject(element);
+        }
+    }
+
+    readonly [Symbol.toStringTag]: string = this._values[Symbol.toStringTag];
+
+    get size() : number {
+        return this._values.size;
+    }
+
+    [Symbol.iterator](): IterableIterator<R> {
+        return undefined;
+    }
+
+    add(value: R): this {
+        this._values.add(value);
+        return this;
+    }
+
+    clear(): void {
+        this._values.clear();
+    }
+
+    delete(value: R): boolean {
+        return this._values.delete(value);
+    }
+
+    entries(): IterableIterator<[R, R]> {
+        return this._values.entries();
+    }
+
+    forEach(callbackfn: (value: R, value2: R, set: Set<R>) => void, thisArg?: any): void {
+        this._values.forEach(callbackfn, thisArg);
+    }
+
+    has(value: R): boolean {
+        return this._values.has(value);
+    }
+
+    keys(): IterableIterator<R> {
+        return this._values.keys();
+    }
+
+    values(): IterableIterator<R> {
+        return this._values.values();
+    }
+
+}
+
+/** @internal */
+export const idToConstructor: { [id: number]: new () => Aggregator<any> } = {
+    [AggregatorConstants.COUNT]: CountAggregator,
```

Comment:
Since readData is not called for aggregators, you don't have to register them, but I am OK with having them registered. 

Note that, we have to register CANONICALIZING_SET, as the readData method is called for it

Reply:
I think it is not about readData is called or not. Isn't it related to whether or not the client receives results of type aggregator? I think the only thing here a client can receive is, as you said, the CANONICALIZING_SET type.